dali/operators/image/resize/resize_base.h

```diff
@@ -20,65 +20,93 @@
+#include "dali/operators/image/resize/resampling_attr.h"
 
 namespace dali {
 
-class DLL_PUBLIC ResamplingFilterAttr {
+template <typename Backend>
+class DLL_PUBLIC ResizeBase {
  public:
-  DLL_PUBLIC ResamplingFilterAttr(const OpSpec &spec);
+  explicit ResizeBase(const OpSpec &spec);
+  ~ResizeBase();
 
-  /**
-   * Filter used when downscaling
-   */
-  kernels::FilterDesc min_filter_{ kernels::ResamplingFilterType::Triangular, 0 };
-  /**
-   * Filter used when upscaling
-   */
-  kernels::FilterDesc mag_filter_{ kernels::ResamplingFilterType::Linear, 0 };
-  /**
-   * Initial size, in bytes, of a temporary buffer for resampling.
-   */
-  size_t temp_buffer_hint_ = 0;
-};
+  void InitializeCPU(int num_threads);
+  void InitializeGPU(int minibatch_size);
 
-class DLL_PUBLIC ResizeBase : public ResamplingFilterAttr {
- public:
-  explicit inline ResizeBase(const OpSpec &spec) : ResamplingFilterAttr(spec) {}
+  using Workspace = workspace_t<Backend>;
 
-  DLL_PUBLIC void Initialize(int num_threads = 1);
-  DLL_PUBLIC void InitializeGPU(int batch_size, int minibatch_size);
+  using input_type =  typename Workspace::template input_t<Backend>::element_type;
+  using output_type = typename Workspace::template output_t<Backend>::element_type;
 
-  DLL_PUBLIC void RunGPU(TensorList<GPUBackend> &output,
-                         const TensorList<GPUBackend> &input,
-                         cudaStream_t stream);
+  void RunResize(Workspace &ws, output_type &output, const input_type &input);
 
-  DLL_PUBLIC void RunCPU(Tensor<CPUBackend> &output,
-                         const Tensor<CPUBackend> &input,
-                         int thread_idx);
+  /**
+   * @param ws                workspace object
+   * @param out_shape         output shape, determined by params
+   * @param input             input; data is not accessed, only shape and metadata are relevant
+   * @param params            resampling parameters; this is a flattened array of size
+   *                          `spatial_ndim*num_samples`, each sample is described by spatial_ndim
+   *                          ResamplingParams, starting from outermost spatial dimension
+   *                          (i.e. [depthwise,] vertical, horizontal)
+   * @param out_type          desired output type
+   * @param first_spatial_dim index of the first resized dim
+   * @param spatial_ndim      number of resized dimensions - these need to form a
+   *                          contiguous block in th layout
+   */
+  void SetupResize(TensorListShape<> &out_shape,
+                   DALIDataType out_type,
+                   const TensorListShape<> &in_shape,
+                   DALIDataType in_type,
+                   span<const kernels::ResamplingParams> params,
+                   int spatial_ndim,
+                   int first_spatial_dim = 0);
 
-  std::vector<kernels::ResamplingParams2D> resample_params_;
-  TensorListShape<> out_shape_;
+  template <size_t spatial_ndim>
```

Comment:
```suggestion
  template <int spatial_ndim>
```
We agreed some time ago to use `int` for template arguments. Any special reason to do otherwise here?

Reply:
Yes - this is a necessity, because ResamplingParamsND is a typedef sfor std::array.